src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,973 @@
+                                        (from principal) 
+                                        (to principal))
+  (if (is-eq from to)
+    (ok true)
+    (begin
+      (unwrap!
+        (nft-transfer? names (tuple (name name) (namespace namespace)) from to)
+        (err ERR_NAME_COULD_NOT_BE_TRANSFERED))
+      (map-delete owner-name ((owner from)))
+      (map-set owner-name
+        ((owner to))
+        ((namespace namespace) (name name)))
+      (ok true))))
+
+(define-private (update-zonefile-and-props (namespace (buff 19))
+                                           (name (buff 16))
+                                           (registered-at (optional uint)) 
+                                           (imported-at (optional uint)) 
+                                           (revoked-at (optional uint)) 
+                                           (zonefile-hash (buff 20)))
+  (let 
+    ((current-page (var-get attachments-inv-page-cursor))
+    (current-index (var-get attachments-inv-index-cursor)))
+    (let 
+      ((next-page (if (is-eq (+ current-index u1) ATTACHMENTS_INV_PAGE_SIZE)
+        (+ current-page u1)
+        current-page))
+      (next-index (mod (+ current-index u1) ATTACHMENTS_INV_PAGE_SIZE)))
+      ;; Emit event used as a system hinter
+      (print {
+        attachment: {
+          hash: zonefile-hash,
+          page-index: current-page,
+          position-in-page: current-index,
+          metadata: {
+            name: name,
+            namespace: namespace,
+            tx-sender: tx-sender
+          }
+        }})
+      ;; Update attachments-inv
+      (map-set attachments-inv
+        ((page current-page) (index current-index))
+        ((content-hash zonefile-hash)))
+      ;; Update cursors
+      (var-set attachments-inv-page-cursor next-page)
+      (var-set attachments-inv-index-cursor next-index)
+      (map-set name-properties
+        ((namespace namespace) (name name))
+        ((registered-at registered-at)
+          (imported-at imported-at)
+          (revoked-at revoked-at)
+          (zonefile-hash zonefile-hash))))))
+
+(define-read-only (get-attachments-inv-info)
+  (ok { 
+    pages-count: (+ (var-get attachments-inv-page-cursor) u1),
+    last-page-len: (var-get attachments-inv-index-cursor),
+    page-size: ATTACHMENTS_INV_PAGE_SIZE
+  }))
+
+;;;; NAMESPACES
+;; NAMESPACE_PREORDER
+;; This step registers the salted hash of the namespace with BNS nodes, and burns the requisite amount of cryptocurrency.
+;; Additionally, this step proves to the BNS nodes that user has honored the BNS consensus rules by including a recent
+;; consensus hash in the transaction.
+;; Returns pre-order's expiration date (in blocks).
+(define-public (namespace-preorder (hashed-salted-namespace (buff 20))
+                                   (stx-to-burn uint))
+  (let 
+    ((former-preorder 
+      (map-get? namespace-preorders ((hashed-salted-namespace hashed-salted-namespace) (buyer contract-caller)))))
```

Comment:
`contract-caller` or `tx-sender`?  I think you may have meant to use the latter?

Reply:
hmm so I was initially thinking about basing everything on `contract-caller`, but I think that @kantai is pretty much sold on using `tx-sender`.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/3e94f3addbe7c6796aba2393a272259199235faf. I really think that we should have a deep debate with strong guidelines on `tx-sender` vs `contract-caller`